src/aof.c

```diff
@@ -40,189 +40,692 @@
-    listNode *ln;
-    aofrwblock *block;
-    ssize_t nwritten;
-    mstime_t latency;
-    UNUSED(el);
-    UNUSED(fd);
-    UNUSED(privdata);
-    UNUSED(mask);
+/*  Load the manifest information from the disk to `server.aof_manifest` 
+ *  when the redis server start.
+ *  
+ *  During the loading process, we will conduct strict error checking. 
+ *  Once there are file opening error, format error, etc., we will 
+ *  directly exit the redis process.
+ * 
+ *  Note: We will ignore the "doesn't exist" error, because this will 
+ *  happen when we upgrade from an old version redis.
+ */
+void aofLoadManifestFromDisk(void) {
+    const char *err = NULL;
+    struct redis_stat sb;
+    long long maxseq = 0;
 
-    latencyStartMonitor(latency);
-    while(1) {
-        ln = listFirst(server.aof_rewrite_buf_blocks);
-        block = ln ? ln->value : NULL;
-        if (server.aof_stop_sending_diff || !block) {
-            aeDeleteFileEvent(server.el,server.aof_pipe_write_data_to_child,
-                              AE_WRITABLE);
-            break;
+    server.aof_manifest = aofManifestCreate();
+
+    sds am_name = getAofManifestFileName();
+    FILE *fp = fopen(am_name, "r");
+    if (fp == NULL) {
+        int en = errno;
+        if (redis_stat(am_name, &sb) == 0) {
+            serverLog(LL_WARNING, "Fatal error: can't open the AOF manifest  \
+                file %s for reading: %s", am_name, strerror(en));
+
+            exit(1);
+        } else {
+            serverLog(LL_NOTICE, "The AOF manifest file %s doesn't exist: %s",
+                am_name, strerror(errno));
+
+            sdsfree(am_name);
+            return;
         }
-        if (block->used != block->pos) {
-            nwritten = write(server.aof_pipe_write_data_to_child,
-                             block->buf+block->pos,block->used-block->pos);
-            if (nwritten <= 0) break;
-            block->pos += nwritten;
+    }  
+
+    sdsfree(am_name);
+
+    char buf[1024];
+    sds *argv = NULL;
+    int argc;
+    aofInfo *ai = NULL;
+
+    sds line;
+    int linenum = 0;
+
+    while (1) {
+        if (fgets(buf, sizeof(buf), fp) == NULL) {
+            if (feof(fp)) {
+                if (linenum == 0) {
+                    err = "Found an empty AOF manifest";
+                    goto loaderr;
+                } else {
+                    break;
+                }
+            } else {
+                err = "Read AOF manifest failed";
+                goto loaderr;
+            }
         }
-        if (block->used == block->pos) listDelNode(server.aof_rewrite_buf_blocks,ln);
+
+        linenum++;
+
+        line = sdsnew(buf);
+        line = sdstrim(line, " \t\r\n");
+
+        argv = sdssplitargs(line, &argc);
+        if (argv == NULL || argc != 6) {
+            err = "The AOF manifest file is invalid format";
+            goto loaderr;
+        }
+
+        ai = aofInfoCreate();
+        if (strcmp(argv[0], AOF_MANIFEST_KEY_FILE_NAME) == 0) {
```

Comment:
I may not understand your question too well, but I think the manifest is a meta file of AOF, which requires a strict format restriction. And the manifest is different from `redis.conf`. `redis.conf` allows users to modify and customize by themselves, but in theory the manifest file can only be generated and modified by redis, so the manifest does not need flexibility, it needs a definite format. I don't know if my answer can solve your question, and I look forward to your feedback.

Reply:
I think Yossi meant that if the manifest lines look like: 
```
file <filename> seq <sequence> type <type> 
```
it implies that they're pairs of key and value, and can be ordered differently, in which case the parser can iterate on the pairs in an if-else chain inside a while loop and don't care about the order.

if it is mandatory that the order is fixed, then we may not need the "key names", and just do
```
<file> <sequence> <type>
```

i.e. regardless of the fact users aren't expected to edit it. the fact we chose this flexible format, means we better also have a flexible parser.

Reply:
Yes, the manifest format is fixed, I think we do not allow the following situations:
```
file appendonly.aof_1.incr.aof seq 1 type i
seq 2 type i file appendonly.aof_2.incr.aof
type i file appendonly.aof_3.incr.aof seq 3
```
The reason why it is displayed in the form of key-value (not just value) is because the user may often look at this file (after all, it is plain text and readable), and the key can clearly tell user the meaning of value. Imagine that when we add other information to the manifest in the future, having a strict key identification can make it easier for us to do format verification. And, in the manifest file, i think we should not lose readability by skimping on a few bytes.

Reply:
i agree, but changing the parsing code to use a while with if-else chain will not complicate the parser much, and will make it flexible.

Reply:
I have modified it to use for loop, Plz review this part  again, thanks.

Reply:
LGTM.
maybe add a comment next to the last `else` and the `<6` to note that it was done for forward compatibility.